### PR TITLE
Keras-like API for training

### PR DESCRIPTION
**Pull request by hkvision**
## What changes were proposed in this pull request?

First version, subject to modifications.
Like keras,
model.compile
model.fit
cc @jason-dai @zhichao-li 

## How was this patch tested?

Jenkins

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 1]
```diff
@@ -816,5 +816,47 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+    // TODO: shape inference
+    this.optimMethod = optimizer
+    this.criterion = loss
+    this.vMethods = metrics
+  }
+
+  /**
+   * Trains the model for a fixed number of epochs.
+   */
+  def fit(x: RDD[Sample[T]], batchSize: Int = 32, epochs: Int = 10,
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Need to check if the model has been compiled

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 2]
```diff
@@ -816,5 +816,47 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
   def getClassTagNumerics() : (Array[ClassTag[_]], Array[TensorNumeric[_]]) = {
     (Array(scala.reflect.classTag[T]), Array(ev))
   }
+
+  protected var optimMethod: OptimMethod[T] = null
+  protected var criterion: Criterion[T] = null
+  protected var vMethods: Array[ValidationMethod[T]] = null
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Maybe add one single object that contains all these variables?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 3]
```diff
@@ -816,5 +816,47 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+
+  /**
+   * Configures the learning process.
+   * Must call this before fit.
+   */
+  def compile(optimizer: OptimMethod[T], loss: Criterion[T],
+              metrics: Array[ValidationMethod[T]] = null): Unit = {
+    // TODO: shape inference
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Can we compile the model multiple times?

Comment (hkvision) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Currently the behavior is that if compiled multiple times, the last compile configurations will take effect.

**Comment by zhichao-li** [comment]
I'm thinking maybe this keras-like API should be with the keras.Model or keras.Sequential instead of mix with the original `AbstractModule`.

**Comment by hkvision** [comment]
@zhichao-li Yeah. After adding more layers, I will remove it into keras packages. cc @jason-dai

**Comment by hkvision** [comment]
Opened a new one. Closed this one.

**Pull request closed without merging**